### ISSUE TITLE
LinearProblemCache (and temp Function Cache) are somewhat icky

### ISSUE THREAD

**Issue by wence-** [thetisproject/thetis#5]
I think I understand why these are here, however, they can cause some slightly strange side effects.

I think the function cache is easier to address.  Effectively you're using this for "work vectors".  When you want a temporary function, you ask for one given a particular function space.

I think we could add something like this to Firedrake.  Here's how I would do it.

The FunctionSpace object gains a `get_function(name=None)`, `restore_function(function)` pair.  You can then get a temporary function that you can use:

```python
f = V.get_function()
... do stuff with f
V.restore_function(f)
```

We can wrap this up in a context manger so one could do:

```python
with V.work_space() as f:
    ... do stuff with f
```

This is neater, because if the only reference to the work space is inside V, once V gets collected, so do the temporary functions.

The linear problem cache is a little trickier, however, I think it is still possible to fix.  The entries in the problem cache become invalid when the mesh they were built on top of go out of scope.  So we could make these collectable by adding a level on top using a `WeakKeyDictionary`.  Sketch:

```python
import weakref
class ProblemCache(object):

    def __init__(self):
        self._cache = weakref.WeakKeyDictionary()

    def __getitem__(self, key):
        mesh, key = key
        return self._cache[mesh][key]

    def __setitem__(self, key, val):
        mesh, key = key
        d = self._cache.get(mesh, {})
        d[key] = val
        self._cache[mesh] = d

    def __contains__(self, key):
        mesh, key = key
        return mesh in self._cache and self._cache[mesh].get(key) is not None

    def clear(self):
        self._cache.clear()
```

So the idea is to use a weakref to the Mesh as a first-level key, the second level acts as before.  Now if the Mesh goes out of scope, the cached problems are no longer strongly referenced by the cache and can be collected.

**Comment by wence-** [comment]
Humph, that doesn't work I think, because the value in the dict holds a strong reference to the key and so it's never collected.

**Comment by tkarna** [comment]
I agree, both of these dicts were implemented as temporary fixes to issues I had.

I use temporary functions essentially to save memory. I think storing them in the function space directly is a great idea. Why is `restore_function` necessary?

**Comment by tkarna** [comment]
As for the LinearProblemCache, I think instead of caching them, another solution would be to implement classes that hold the solver objects. For functions like `computeVertVelocity` that certainly makes sense. So essentially all those functions in `utility` would be implemented as objects that are stored in the solver class:

```python
solverObj.uv_vert_integrator = ComputeVerticalIntegral(uv_3d, uv_dav_3d, average=True)
solverObj.uv_vert_integrator.apply(uv_3d, uv_dav_3d)
```

That should take care of the collection issue.

**Comment by wence-** [comment]
So I had thought the restore_function would be used like this:

```python

tmp = V.get_function()
... do stuff with tmp
tmp2 = V.get_function()
... do stuff with tmp2
V.restore_function(tmp)
... do stuff
tmp3 = V.get_function()
# tmp3 doesn't allocate a new work vector, instead it gets the work vector
# that was previously tmp.
```

IOW, if I only ever need one temporary function on `V` but I use it repeatedly in different places, I don't keep on allocating new functions (or allocating and deallocating).  But rather `V` has a (fixed size) number of work vectors that I can grab hold of and then release back to the pool when I'm done with them.

**Comment by wence-** [comment]
That approach to the linear problem cache also sounds sensible.

**Comment by tkarna** [comment]
OK, I see. Yes, having a pool of temporary functions seems like a good approach.
I'll start implementing those changes in `utility.py` so we can remove `LinearProblemCache`.

**Comment by tkarna** [comment]
linProblemCache is now removed in pull request #8 .

**Comment by wence-** [comment]
Please look at firedrake branch "work_functions" and see if that fits the bill for the temporary function cache.

**Comment by tkarna** [comment]
OK, looks good, I'll give it a shot.

**Comment by tkarna** [comment]
So I think it's possible to use the function space work functions. The only caveat is that during initialization we build Projector objects with the work functions and then re-use those Projectors at run time. This requires that you always get the same work function. I've worked around this by setting max_work_functions to 1, which is fine for now but might be too restrictive in some cases.

**Comment by tkarna** [comment]
Both issues are now addressed in pull requests #8 and #21, closing.
